src/Arrays/MPIStateArrays.jl

```diff
@@ -353,39 +300,23 @@ the listed states are summed, otherwise all the states in `A` are used.
 
 A typical use case for this is when the weights have been initialized with
 quadrature weights from a grid, thus this becomes an integral approximation.
-
-!!! note
-
-    This implementation is not optimal and should be revisited when we work out
-    the GPUify version!
-
 """
 function weightedsum(A::MPIStateArray, states=1:size(A, 2))
-
-  host_array = Array ∈ typeof(A).parameters
-  h_A = host_array ? A : Array(A)
-  Np = size(A, 1)
-
   isempty(A.weights) && error("`weightedsum` requires weights")
-  locwsum = knl_weightedsum(Val(Np), h_A, A.weights, A.realelems, states)
 
-  DFloat = eltype(A)
-  # Need to use anomous function version of sum else MPI.jl using MPI_SUM
-  DFloat(MPI.Allreduce([locwsum], (x,y)->x+y, A.mpicomm)[1])
-end
+  T = eltype(A)
+  states = SVector{length(states)}(states)
 
-function knl_weightedsum(::Val{Np}, A, weights, elems, states) where {Np}
-  DFloat = eltype(A)
+  C = @view A.Q[:, states, A.realelems]
+  w = @view A.weights[:, :, A.realelems]
+  init = zero(DoubleFloat{T})
 
-  wsum = zero(DoubleFloat{DFloat})
+  E = @~ DoubleFloat{T}.(C) .* DoubleFloat{T}.(w)
 
-  @inbounds for e = elems
-    for q = states, i = 1:Np
-      wsum += weights[i, e] * A[i, q, e]
-    end
-  end
+  locwsum = mapreduce(identity, +, E, init=init)
 
-  wsum
+  # Need to use anomous function version of sum else MPI.jl using MPI_SUM
+  T(MPI.Allreduce([locwsum], (x,y)->x+y, A.mpicomm)[1])
```

Comment:
why can't you use `MPI_SUM`?

Reply:
That was there before so I kept it.  I believe `MPI.jl` was selecting `MPI_SUM` for variables of type `DoubleFloat{DFloat}` which doesn't work.  However, I didn't test it.

Reply:
Ah right